### PR TITLE
[x] add lint command

### PR DESCRIPTION
**Pull request by bmwill**
Introduce a `lint` command to x. `lint` is the begining of a general
    purpose linting engine for libra which we can use to write custom lints
    which may not be covered by clippy and which we may want to run on
    non-rust files.

### CODE REVIEW

**Comment by bmwill** [comment]
I need to do a check on the proto files too for the license checking lint

File: x/src/lint.rs [hunk 1]
```diff
@@ -0,0 +1,164 @@
+use crate::{config::Config, utils::project_root};
```
Comment (davidiw) [inline x/src/lint.rs]: already violated the lint in the first diff, no licensing :P

File: x/src/lint.rs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+use anyhow::anyhow;
```
Comment (davidiw) [inline x/src/lint.rs]: would really love to see standardized error handling in our code :O

File: x/src/lint.rs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    let mut full_file_path = PathBuf::new();
+
+    for file in get_file_listing()? {
+        full_file_path.push(&project_root);
+        full_file_path.push(&file);
+
+        let contents = match fs::read_to_string(&full_file_path) {
+            Ok(s) => s,
+            // Skip non utf8 files
+            Err(_) => continue,
```
Comment (davidiw) [inline x/src/lint.rs]: what if the file ends in .rs, .mvir, or .sh? Should this be a lint violation?

File: x/src/lint.rs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+                               SPDX-License-Identifier: Apache-2.0\n\
+                               ";
+
+fn has_license_header(file: &Path, contents: &str) -> Result<(), String> {
+    enum FileType {
+        Rust,
+        Shell,
+        MoveIr,
+        Proto,
+    }
```
Comment (davidiw) [inline x/src/lint.rs]: why not

```
enum FileType {
    Rust = "rs",
...
}
```

Comment (davidiw) [reply x/src/lint.rs]: Perhaps we still need the boiler plate code anyway... just curious to hear your thoughts.

File: docker/circleci/build.sh [hunk 5]
```diff
@@ -1,4 +1,6 @@
 #!/bin/sh
+# Copyright (c) The Libra Core Contributors
+# SPDX-License-Identifier: Apache-2.0
 set -e
```
Comment (davidiw) [inline docker/circleci/build.sh]: maybe the first line should be empty?

File: x/src/lint.rs (hunk 4 above)
Comment (bmwill) [reply x/src/lint.rs]: you can't do this, only integer types are accepted with that style of enum.

File: x/src/lint.rs (hunk 2 above)
Comment (bmwill) [reply x/src/lint.rs]: I would too, the current crate we're using, failure, is not what we should be using long term.

File: x/src/lint.rs (hunk 1 above)
Comment (bmwill) [reply x/src/lint.rs]: just proving my point :)

**Comment by bmwill** [comment]
Good point and breaking up the lint driver from the lint definitions, i'll do that before landing. As for the license in the shell files that's more of a style/formatting thing I don't feel too strongly about it but i don't think it should be part of the license lint.

**Pull request closed without merging**